src/net_processing.cpp

```diff
@@ -2259,10 +2258,18 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
     }
 
     BlockValidationState state;
+    const CBlockIndex* pindexLast{nullptr};
     if (!m_chainman.ProcessNewBlockHeaders(headers, state, &pindexLast)) {
         if (state.IsInvalid()) {
             MaybePunishNodeForBlock(pfrom.GetId(), state, via_compact_block, "invalid header received");
-            return;
+
+            // The received headers chain could contain valid headers followed
+            // by invalid headers. We still want to continue processing the
+            // valid headers. `pindexLast` will point to block index of the
+            // last valid header.
+            if (!pindexLast) {
+                return;
```

Comment:
Hmm, so this is a bigger change than I was trying to suggest; I only meant that we should update the chain tip set if pindexLast is non null, even if an invalid block is encountered.  I'm not sure we should invoke all the logic that follows, because some of the processing after this includes sending new p2p messages (such as getheaders if the message was full), which wouldn't make sense if we encountered an invalid header, even if we weren't going to disconnect the peer.

I think it's possible that some of the logic that follows _could_ be invoked for the valid headers if we knew that we we weren't about to disconnect this peer, but I feel like such a significant behavior change is out of scope for this PR.  

My suggestion would be to drop this commit and instead just invoke `UpdateChainTipSet` for `pindexLast` (if non-null) before returning, after calling `MaybePunishNodeForBlock`?

Reply:
You are right, i went a bit overboard with this. Changed it to just update the chain tip set like you suggested.